libraries/ESP8266WiFi/src/WiFiClientBearSSL.cpp

```diff
@@ -0,0 +1,1203 @@
+/*
+  WiFiClientBearSSL- SSL client/server for esp8266 using BearSSL libraries
+  - Mostly compatible with Arduino WiFi shield library and standard
+    WiFiClient/ServerSecure (except for certificate handling).
+
+  Copyright (c) 2018 Earle F. Philhower, III
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#define LWIP_INTERNAL
+
+extern "C"
+{
+#include "osapi.h"
+#include "ets_sys.h"
+}
+#include <list>
+#include <errno.h>
+#include "debug.h"
+#include "ESP8266WiFi.h"
+#include "WiFiClient.h"
+#include "WiFiClientBearSSL.h"
+#include "lwip/opt.h"
+#include "lwip/ip.h"
+#include "lwip/tcp.h"
+#include "lwip/inet.h"
+#include "lwip/netif.h"
+#include "include/ClientContext.h"
+#include "c_types.h"
+
+// BearSSL needs a very large stack, larger than the entire ESP8266 Arduino
+// default one.  This shared_pointer is allocated on first use and cleared
+// on last cleanup, with only one stack no matter how many SSL objects.
+std::shared_ptr<uint8_t> WiFiClientBearSSL::_bearssl_stack = nullptr;
+
+
+
+void WiFiClientBearSSL::_clear()
+{
+  // TLS handshake may take more than the 5 second default timeout
+  _timeout = 15000;
+
+  _sc = nullptr;
+  _sc_svr = nullptr;
+  _eng = nullptr;
+  _x509_minimal = nullptr;
+  _x509_insecure = nullptr;
+  _x509_knownkey = nullptr;
+  _iobuf_in = nullptr;
+  _iobuf_out = nullptr;
+  _now = 0; // You can override or ensure time() is correct w/configTime
+  _ta = nullptr;
+  setBufferSizes(16384, 512); // Minimum safe
+  _handshake_done = false;
+  _recvapp_buf = nullptr;
+  _recvapp_len = 0;
+  _oom_err = false;
+}
+
+void WiFiClientBearSSL::_clearAuthenticationSettings()
+{
+  _use_insecure = false;
+  _use_fingerprint = false;
+  _use_self_signed = false;
+  _knownkey = nullptr;
+  _sk = nullptr;
+}
+
+
+WiFiClientBearSSL::WiFiClientBearSSL() : WiFiClient()
+{
+  _clear();
+  _clearAuthenticationSettings();
+  _certStore = nullptr; // Don't want to remove cert store on a clear, should be long lived
+  if (!_bearssl_stack) {
+#if 1
+    const int stacksize = 4500; // Empirically determined stack for EC and RSA connections
+    _bearssl_stack = std::shared_ptr<uint8_t>(new uint8_t[stacksize], std::default_delete<uint8_t[]>());
+    br_esp8266_stack_proxy_init(_bearssl_stack.get(), stacksize);
+#else
+    // TODO - DANGEROUS - EVIL
```

Comment:
Will this be done/tested?

Reply:
I've tested moving the 2ndary stack to this unused spot and it runs fine though everything I've thrown at it.  Extra 4.5KB free, no issues.  However I'm not sure if we want to unilaterally grab it here.

Reply:
No actually the current EC crypto contains yield, so we can no longer use the sys context freely. 2ndary stack seems to be the only viable option at this moment.

Reply:
@Adam5Wu Not sure I understand your comment.  The code right now has the structure:
````
#if 1
  _fake_stack = <malloc space>
#else
 _fake_stack = <hardcoded address beginning at unused space in SYS region>
#endif
````
Are you saying we can't use anything in the SYS region (via the 2ndary stack code changes, not via changing the SP) if there's a yield?

I don't think I've seen an issue, even with EC and the yield().  The MFLN example which hits the EC and definitely needs to yield (thanks to tls.mbed.org's handshake) runs like a champ borrowing that space...

Reply:
I was actually thinking of doing some magic with the SP itself and removing the STACK_PROXY magic in bearssl libs.  With careful thinking back and forth (and wrapping the optimistic_yield() call in thunks as well) I think it could work.  But it's an optimization I didn't have bandwidth to try out. and it would confuse the stack dumper to no end on crashes.

Reply:
According to:
https://github.com/esp8266/Arduino/blob/42f824b2e4a1df8812306d40f2ea59a1323618c1/cores/esp8266/cont.S#L50-L59

When the current context is yield, stack pointer is restored to point into the sys cont area (overlaps with "evil“ crypto 2ndary stack), which means data in there may be overwritten when the SDK code runs the next.

Reply:
Ah, I see.  That said, though, because the code was stealing it even w/o the yield(), the SYS stack could have grown to that unused space anyway.  We'd cause a very bad crash sometime later when the SYS code tried to pop things we'd overwritten.

If we want to steal it then I think we need some way of causing max SYS stack usage and grabbing it (maybe after the fact by filling this space w/canaries and checking for the first one overwritten).

The proper solution, of course, is for the binary layer to not allocate that space at all and move everything up to give us those 5KB of RAM as heap...

Reply:
If you really want to get rid of stack proxy, I think it may be possible (and more "proper" in the eyes of stack dumper) to create a new context with a bigger stack and run the heavy crypto in there.
https://github.com/esp8266/Arduino/blob/42f824b2e4a1df8812306d40f2ea59a1323618c1/cores/esp8266/core_esp8266_main.cpp#L159
https://github.com/esp8266/Arduino/blob/42f824b2e4a1df8812306d40f2ea59a1323618c1/cores/esp8266/core_esp8266_main.cpp#L129

But I kind of like the current stack proxy, I think the code is clever and easier to work with than making a new cont. :)

Reply:
@Adam5Wu, you're a genius.

The **best** thing to do with that extra "unused" space is to **move g_cont** into it via the linker.  Every app gets to benefit from the 4+KB heap savings, and BearSSL will just use the STACK_PROXY on the heap or a build a new context on the heap (why didn't I think of using a ctx...I was experimenting with asm() blocks...) 

STACK_PROXY seems to work well now, but I am a little worried about pulling in @pornin's changes when he releases new BearSSL code.  Diffs will need to be scrubbed and the appropriate ENTRY and EXITs added.

Reply:
> I am a little worried about pulling in @pornin's changes when he releases new BearSSL code.
Yeap I have similar concerns, and I have attempted to convert the proxies variable into a form more closer to the original source, but it was unsuccessful due to C lack of reference type... :P